src/VscPreviewExtension/src/extension.ts

```diff
@@ -2,44 +2,96 @@
+const port = 8080;
+
+let startLine = 0;
+let endLine = 0;
 
 export function activate(context: ExtensionContext) {
     let dfmProcess = new PreviewCore(context);
-    let providerRegistration = workspace.registerTextDocumentContentProvider("markdown", dfmProcess.provider);
+    let tokenTreeProcess = new TokenTreeCore(context, port);
+    let previewProviderRegistration = workspace.registerTextDocumentContentProvider(markdownScheme, dfmProcess.provider);
+    let tokenTreeProviderRegistration = workspace.registerTextDocumentContentProvider(tokenTreeScheme, tokenTreeProcess.provider);
 
     // Event register
     let showPreviewRegistration = commands.registerCommand("DFM.showPreview", uri => showPreview(dfmProcess));
     let showPreviewToSideRegistration = commands.registerCommand("DFM.showPreviewToSide", uri => showPreview(dfmProcess, uri, true));
     let showSourceRegistration = commands.registerCommand("DFM.showSource", showSource);
+    let showTokenTreeToSideRegistration = commands.registerCommand("DFM.showTokenTreeToSide", uri => showTokenTree(tokenTreeProcess));
 
-    context.subscriptions.push(showPreviewRegistration, showPreviewToSideRegistration, showSourceRegistration, providerRegistration);
+    context.subscriptions.push(showPreviewRegistration, showPreviewToSideRegistration, showSourceRegistration, showTokenTreeToSideRegistration);
+    context.subscriptions.push(previewProviderRegistration, tokenTreeProviderRegistration);
 
     workspace.onDidSaveTextDocument(document => {
         if (isMarkdownFile(document)) {
             const uri = getMarkdownUri(document.uri);
             dfmProcess.callDfm(uri);
+
+            tokenTreeProcess.callDfm(getTokenTreeUri(document.uri));
         }
     });
 
     workspace.onDidChangeTextDocument(event => {
         if (isMarkdownFile(event.document)) {
             const uri = getMarkdownUri(event.document.uri);
             dfmProcess.callDfm(uri);
+            // TODO: make token tree change synchronous
+            // tokenTreeProcess.callDfm(getTokenTreeUri(event.document.uri));
         }
     });
 
     workspace.onDidChangeConfiguration(() => {
         workspace.textDocuments.forEach(document => {
             if (document.uri.scheme === markdownScheme) {
                 dfmProcess.callDfm(document.uri);
+            } else if (document.uri.scheme === tokenTreeScheme) {
+                tokenTreeProcess.callDfm(document.uri);
             }
         });
     });
+
+    window.onDidChangeTextEditorSelection(event => {
+        startLine = event.selections[0].start.line + 1;
+        endLine = event.selections[0].end.line + 1;
+    });
+
+    // Http server to communicate with js
+    let http = require("http");
+    let server = http.createServer();
+    server.on("request", function (req, res) {
+        let requestInfo = req.url.split("/");
+        if (requestInfo[1] === "lineNumber") {
+            mapToSelection(parseInt(requestInfo[2]), parseInt(requestInfo[3]));
```

Comment:
>mapToSelection [](start = 12, length = 14)

could make `mapToSelection` return true if it works successful, otherwise return false and show error message?